Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp

```diff
@@ -0,0 +1,310 @@
+    }
+
+    g_signal_connect(m_proxy.get(), "g-signal", G_CALLBACK(handleSignal), this);
+}
+
+void NotificationService::processCapabilities(GVariant* variant)
+{
+    GUniqueOutPtr<GVariantIter> iter;
+    g_variant_get(variant, "(as)", &iter.outPtr());
+    const char* capability;
+    while (g_variant_iter_loop(iter.get(), "&s", &capability)) {
+        if (!g_strcmp0(capability, "action-icons"))
+            m_capabilities.add(Capabilities::ActionIcons);
+        else if (!g_strcmp0(capability, "actions"))
+            m_capabilities.add(Capabilities::Actions);
+        else if (!g_strcmp0(capability, "body"))
+            m_capabilities.add(Capabilities::Body);
+        else if (!g_strcmp0(capability, "body-hyperlinks"))
+            m_capabilities.add(Capabilities::BodyHyperlinks);
+        else if (!g_strcmp0(capability, "body-images"))
+            m_capabilities.add(Capabilities::BodyImages);
+        else if (!g_strcmp0(capability, "body-markup"))
+            m_capabilities.add(Capabilities::BodyMarkup);
+        else if (!g_strcmp0(capability, "icon-multi"))
+            m_capabilities.add(Capabilities::IconMulti);
+        else if (!g_strcmp0(capability, "icon-static"))
+            m_capabilities.add(Capabilities::IconStatic);
+        else if (!g_strcmp0(capability, "persistence"))
+            m_capabilities.add(Capabilities::Persistence);
+        else if (!g_strcmp0(capability, "sound"))
+            m_capabilities.add(Capabilities::Sound);
+    }
+}
+
+bool NotificationService::showNotification(const WebNotification& notification)
+{
+    if (!m_proxy)
+        return false;
+
+    auto findNotificationByTag = [&](const String& tag) -> Notification {
+        if (tag.isEmpty())
+            return Notification();
+
+        uint64_t notificationID = 0;
+        for (const auto& it : m_notifications) {
+            if (it.value.tag == tag) {
+                notificationID = it.key;
+                break;
+            }
+        }
+
+        return notificationID ? m_notifications.take(notificationID) : Notification({ 0, { }, tag });
+    };
+
+    auto addResult = m_notifications.add(notification.notificationID(), findNotificationByTag(notification.tag()));
+
+    if (shouldUsePortal()) {
+        GVariantBuilder builder;
+        g_variant_builder_init(&builder, G_VARIANT_TYPE_VARDICT);
+
+        g_variant_builder_add(&builder, "{sv}", "title", g_variant_new_string(notification.title().utf8().data()));
+        g_variant_builder_add(&builder, "{sv}", "body", g_variant_new_string(notification.body().utf8().data()));
+        g_variant_builder_add(&builder, "{sv}", "default-action", g_variant_new_string("default"));
+        addResult.iterator->value.portalID = createVersion4UUIDString();
+        g_dbus_proxy_call(m_proxy.get(), "AddNotification", g_variant_new("(s@a{sv})", addResult.iterator->value.portalID.utf8().data(), g_variant_builder_end(&builder)),
+            G_DBUS_CALL_FLAGS_NONE, -1, nullptr, nullptr, nullptr);
```

Comment:
Should pass a callback just to print a warning if it fails. Yes, it's slightly more effort, but the future benefit in case things break in the future will be worth it.

Reply:
Ok